src/XR/features/WebXRHandTracking.ts

```diff
@@ -324,112 +367,196 @@ export class WebXRHand implements IDisposable {
+     * Update this hand from the latest xr frame.
+     * @param xrFrame The latest frame received from WebXR.
+     * @param referenceSpace The current viewer reference space.
      */
-    public updateFromXRFrame(xrFrame: XRFrame, referenceSpace: XRReferenceSpace, scaleFactor: number = 2) {
+    public updateFromXRFrame(xrFrame: XRFrame, referenceSpace: XRReferenceSpace) {
         const hand = this.xrController.inputSource.hand;
         if (!hand) {
             return;
         }
-        this.trackedMeshes.forEach((mesh, name) => {
-            const xrJoint = hand.get(name);
-            if (xrJoint) {
-                let pose = xrFrame.getJointPose!(xrJoint, referenceSpace);
-                if (!pose || !pose.transform) {
-                    return;
-                }
-                // get the transformation. can be done with matrix decomposition as well
-                const pos = pose.transform.position;
-                const orientation = pose.transform.orientation;
-                mesh.position.set(pos.x, pos.y, pos.z);
-                mesh.rotationQuaternion!.set(orientation.x, orientation.y, orientation.z, orientation.w);
-                // left handed system conversion
-                // get the radius of the joint. In general it is static, but just in case it does change we update it on each frame.
-                const radius = (pose.radius || 0.008) * scaleFactor;
-                mesh.scaling.set(radius, radius, radius);
-
-                // if we are using left-handed meshes, transform before applying to the hand mesh
-                if (this._leftHandedMeshes && !mesh.getScene().useRightHandedSystem) {
-                    mesh.position.z *= -1;
-                    mesh.rotationQuaternion!.z *= -1;
-                    mesh.rotationQuaternion!.w *= -1;
+
+        const anyHand: any = hand;
+        const jointSpaces: XRJointSpace[] = handJointReferenceArray.map((jointName) => anyHand[jointName]);
+        let trackingSuccessful = true;
+
+        if (xrFrame.fillPoses && xrFrame.fillJointRadii) {
+            trackingSuccessful = xrFrame.fillPoses(jointSpaces, referenceSpace, this._jointTransformMatrices) &&
+                xrFrame.fillJointRadii(jointSpaces, this._jointRadii);
+        } else if (xrFrame.getJointPose) {
+            // Warning: This codepath is slow by comparison, only here for compat.
+            for (let jointIdx = 0; jointIdx < jointSpaces.length; jointIdx++) {
+                const jointPose = xrFrame.getJointPose(jointSpaces[jointIdx], referenceSpace);
+                if (jointPose) {
+                    this._jointTransformMatrices.set(jointPose.transform.matrix, jointIdx * 16);
+                    this._jointRadii[jointIdx] = jointPose.radius || 0.008;
+                } else {
+                    trackingSuccessful = false;
+                    break;
                 }
+            }
+        } else {
+            throw new Error("XRFrame does not support joint tracking!");

```

Comment:
the feature should not be enabled if joint tracking is not supported. is there a case where this will fail even if hands are supported?

Reply:
We leave it up to the user to not enable the feature if joint tracking / hands aren't supported? Doesn't seem like there are any explicit checks against it.

Reply:
The feature "`isCompatible` function should catch that earlier, but it's fine to stay defensive here.